### PR TITLE
Add mypy and typing

### PR DESCRIPTION
**Pull request by bfirsh**
Cog is all about Python types, so why aren't we using it on Cog's codebase?

I ran mypy against Cog for fun and then it complained a lot and then realized this was going to be less fun than I anticipated. So, I am merely flipping the switch here and leaving the rest for us to chip away at over time.

Lots of fun here for those so inclined: 🎹🎹🎹

```
python/cog/server/log_capture.py:9: error: Function is missing a type annotation
python/cog/server/log_capture.py:85: error: Function is missing a type annotation
python/cog/server/log_capture.py:98: error: Function is missing a return type annotation
python/cog/server/log_capture.py:98: note: Use "-> None" if function does not return a value
python/cog/types.py:16: error: Function is missing a return type annotation
python/cog/types.py:16: error: Function is missing a type annotation for one or more arguments
python/cog/types.py:39: error: Function is missing a type annotation
python/cog/types.py:52: error: Function is missing a return type annotation
python/cog/types.py:77: error: Function is missing a type annotation
python/cog/types.py:87: error: Function is missing a return type annotation
python/cog/types.py:101: error: Function is missing a type annotation
python/cog/response.py:13: error: Function is missing a return type annotation
python/cog/predictor.py:157: error: No overload variant of "StringEnum" matches argument types "str", "Dict[Any, Any]"
python/cog/predictor.py:157: note: Possible overload variants:
python/cog/predictor.py:157: note:     def StringEnum(cls, object: object = ...) -> StringEnum
python/cog/predictor.py:157: note:     def StringEnum(cls, object: bytes, encoding: str = ..., errors: str = ...) -> StringEnum
python/cog/predictor.py:159: error: Name "InputType" already defined on line 122
python/cog/predictor.py:159: error: IntEnum() expects a string literal as the first argument
python/cog/json.py:18: error: Function is missing a type annotation for one or more arguments
python/cog/server/runner.py:22: error: Function is missing a return type annotation
python/cog/server/runner.py:22: note: Use "-> None" if function does not return a value
python/cog/server/runner.py:40: error: Function is missing a return type annotation
python/cog/server/runner.py:40: note: Use "-> None" if function does not return a value
python/cog/server/runner.py:57: error: Function is missing a return type annotation
python/cog/server/runner.py:57: note: Use "-> None" if function does not return a value
python/cog/server/runner.py:68: error: Function is missing a type annotation
python/cog/server/runner.py:94: error: Function is missing a return type annotation
python/cog/server/runner.py:108: error: Function is missing a return type annotation
python/cog/server/runner.py:111: error: Function is missing a return type annotation
python/cog/server/runner.py:123: error: Function is missing a return type annotation
python/cog/server/runner.py:126: error: Function is missing a return type annotation
python/cog/server/runner.py:135: error: Function is missing a return type annotation
python/cog/server/runner.py:154: error: Function is missing a type annotation
python/cog/server/runner.py:191: error: Function is missing a return type annotation
python/cog/server/runner.py:205: error: Function is missing a type annotation
python/cog/server/runner.py:217: error: Function is missing a type annotation
python/tests/test_json.py:12: error: Function is missing a return type annotation
python/tests/test_json.py:12: note: Use "-> None" if function does not return a value
python/tests/test_json.py:17: error: Function is missing a return type annotation
python/tests/test_json.py:17: note: Use "-> None" if function does not return a value
python/tests/test_json.py:28: error: Function is missing a return type annotation
python/tests/test_json.py:28: note: Use "-> None" if function does not return a value
python/tests/test_json.py:42: error: Function is missing a return type annotation
python/tests/test_json.py:42: note: Use "-> None" if function does not return a value
python/cog/server/redis_queue.py:25: error: Function is missing a type annotation
python/cog/server/redis_queue.py:32: error: Function is missing a type annotation
python/cog/server/redis_queue.py:35: error: Function is missing a type annotation
python/cog/server/redis_queue.py:43: error: Function is missing a type annotation
python/cog/server/redis_queue.py:95: error: Function is missing a type annotation
python/cog/server/redis_queue.py:99: error: Function is missing a return type annotation
python/cog/server/redis_queue.py:132: error: Function is missing a return type annotation
python/cog/server/redis_queue.py:132: note: Use "-> None" if function does not return a value
python/cog/server/redis_queue.py:187: error: Function is missing a type annotation
python/cog/server/redis_queue.py:280: error: Function is missing a type annotation
python/cog/server/redis_queue.py:285: error: Function is missing a type annotation
python/cog/server/redis_queue.py:295: error: Function is missing a type annotation
python/cog/server/redis_queue.py:304: error: Function is missing a type annotation
python/cog/server/http.py:30: error: Function is missing a return type annotation
python/cog/server/http.py:58: error: Function is missing a return type annotation
python/tests/server/test_http.py:10: error: Skipping analyzing "PIL": module is installed, but missing library stubs or py.typed marker
python/tests/server/test_http.py:10: note: See https://mypy.readthedocs.io/en/stable/running_mypy.html#missing-imports
python/tests/server/test_http.py:18: error: Function is missing a type annotation for one or more arguments
python/tests/server/test_http.py:24: error: Function is missing a return type annotation
python/tests/server/test_http.py:26: error: Function is missing a return type annotation
python/tests/server/test_http.py:26: note: Use "-> None" if function does not return a value
python/tests/server/test_http.py:29: error: Signature of "predict" incompatible with supertype "BasePredictor"
python/tests/server/test_http.py:29: note:      Superclass:
python/tests/server/test_http.py:29: note:          def predict(self, **kwargs: Any) -> Any
python/tests/server/test_http.py:29: note:      Subclass:
python/tests/server/test_http.py:29: note:          def predict(self) -> str
python/tests/server/test_http.py:38: error: Function is missing a return type annotation
python/tests/server/test_http.py:38: note: Use "-> None" if function does not return a value
python/tests/server/test_http.py:40: error: Signature of "predict" incompatible with supertype "BasePredictor"
python/tests/server/test_http.py:40: note:      Superclass:
python/tests/server/test_http.py:40: note:          def predict(self, **kwargs: Any) -> Any
python/tests/server/test_http.py:40: note:      Subclass:
python/tests/server/test_http.py:40: note:          def predict(self, no_default: str, default_without_input: str = ..., input_with_default: int = ..., path: Path = ..., image: File = ..., choices: str = ..., int_choices: int = ...) -> str
python/tests/server/test_http.py:227: error: Function is missing a return type annotation
python/tests/server/test_http.py:227: note: Use "-> None" if function does not return a value
python/tests/server/test_http.py:234: error: Signature of "predict" incompatible with supertype "BasePredictor"
python/tests/server/test_http.py:234: note:      Superclass:
python/tests/server/test_http.py:234: note:          def predict(self, **kwargs: Any) -> Any
python/tests/server/test_http.py:234: note:      Subclass:
python/tests/server/test_http.py:234: note:          def predict(self) -> MyOutput
python/tests/server/test_http.py:374: error: Function is missing a return type annotation
python/tests/server/test_http.py:374: note: Use "-> None" if function does not return a value
python/tests/server/test_http.py:381: error: Signature of "predict" incompatible with supertype "BasePredictor"
python/tests/server/test_http.py:381: note:      Superclass:
python/tests/server/test_http.py:381: note:          def predict(self, **kwargs: Any) -> Any
python/tests/server/test_http.py:381: note:      Subclass:
python/tests/server/test_http.py:381: note:          def predict(self) -> Output
python/tests/server/test_http.py:404: error: Function is missing a return type annotation
python/tests/server/test_http.py:404: note: Use "-> None" if function does not return a value
python/tests/server/test_http.py:406: error: Signature of "predict" incompatible with supertype "BasePredictor"
python/tests/server/test_http.py:406: note:      Superclass:
python/tests/server/test_http.py:406: note:          def predict(self, **kwargs: Any) -> Any
python/tests/server/test_http.py:406: note:      Subclass:
python/tests/server/test_http.py:406: note:          def predict(self) -> Iterator[str]
python/tests/server/test_http.py:424: error: Function is missing a return type annotation
python/tests/server/test_http.py:424: note: Use "-> None" if function does not return a value
python/tests/server/test_http.py:426: error: Signature of "predict" incompatible with supertype "BasePredictor"
python/tests/server/test_http.py:426: note:      Superclass:
python/tests/server/test_http.py:426: note:          def predict(self, **kwargs: Any) -> Any
python/tests/server/test_http.py:426: note:      Subclass:
python/tests/server/test_http.py:426: note:          def predict(self, pick_a_number_any_number: int = ...) -> str
python/tests/server/test_http.py:444: error: Function is missing a return type annotation
python/tests/server/test_http.py:444: note: Use "-> None" if function does not return a value
python/tests/server/test_http.py:446: error: Signature of "predict" incompatible with supertype "BasePredictor"
python/tests/server/test_http.py:446: note:      Superclass:
python/tests/server/test_http.py:446: note:          def predict(self, **kwargs: Any) -> Any
python/tests/server/test_http.py:446: note:      Subclass:
python/tests/server/test_http.py:446: note:          def predict(self) -> Iterator[str]
python/tests/server/test_http.py:457: error: Function is missing a return type annotation
python/tests/server/test_http.py:457: note: Use "-> None" if function does not return a value
python/tests/server/test_http.py:459: error: Signature of "predict" incompatible with supertype "BasePredictor"
python/tests/server/test_http.py:459: note:      Superclass:
python/tests/server/test_http.py:459: note:          def predict(self, **kwargs: Any) -> Any
python/tests/server/test_http.py:459: note:      Subclass:
python/tests/server/test_http.py:459: note:          def predict(self, file: Path) -> Iterator[str]
python/tests/server/test_http.py:478: error: Function is missing a return type annotation
python/tests/server/test_http.py:480: error: Signature of "predict" incompatible with supertype "BasePredictor"
python/tests/server/test_http.py:480: note:      Superclass:
python/tests/server/test_http.py:480: note:          def predict(self, **kwargs: Any) -> Any
python/tests/server/test_http.py:480: note:      Subclass:
python/tests/server/test_http.py:480: note:          def predict(self) -> Iterator[Path]
python/tests/server/test_http.py:495: error: Function is missing a type annotation
python/tests/server/test_http.py:508: error: Function is missing a type annotation
python/tests/server/test_http.py:510: error: Function is missing a return type annotation
python/tests/server/test_http.py:510: note: Use "-> None" if function does not return a value
python/tests/server/test_http.py:513: error: Function is missing a return type annotation
python/tests/server/test_http.py:524: error: Function is missing a return type annotation
python/tests/server/test_http.py:526: error: Function is missing a type annotation for one or more arguments
python/tests/server/test_http.py:526: error: Signature of "predict" incompatible with supertype "BasePredictor"
python/tests/server/test_http.py:526: note:      Superclass:
python/tests/server/test_http.py:526: note:          def predict(self, **kwargs: Any) -> Any
python/tests/server/test_http.py:526: note:      Subclass:
python/tests/server/test_http.py:526: note:          def predict(self, input: Any) -> str
python/tests/server/test_http.py:533: error: Function is missing a return type annotation
python/tests/server/test_http.py:538: error: Signature of "predict" incompatible with supertype "BasePredictor"
python/tests/server/test_http.py:538: note:      Superclass:
python/tests/server/test_http.py:538: note:          def predict(self, **kwargs: Any) -> Any
python/tests/server/test_http.py:538: note:      Subclass:
python/tests/server/test_http.py:538: note:          def predict(self, input: Input) -> str
python/tests/server/test_http_output.py:7: error: Skipping analyzing "PIL": module is installed, but missing library stubs or py.typed marker
python/tests/server/test_http_output.py:15: error: Function is missing a return type annotation
python/tests/server/test_http_output.py:17: error: Signature of "predict" incompatible with supertype "BasePredictor"
python/tests/server/test_http_output.py:17: note:      Superclass:
python/tests/server/test_http_output.py:17: note:          def predict(self, **kwargs: Any) -> Any
python/tests/server/test_http_output.py:17: note:      Subclass:
python/tests/server/test_http_output.py:17: note:          def predict(self) -> int
python/tests/server/test_http_output.py:18: error: Incompatible return value type (got "str", expected "int")
python/tests/server/test_http_output.py:25: error: Function is missing a return type annotation
python/tests/server/test_http_output.py:27: error: Signature of "predict" incompatible with supertype "BasePredictor"
python/tests/server/test_http_output.py:27: note:      Superclass:
python/tests/server/test_http_output.py:27: note:          def predict(self, **kwargs: Any) -> Any
python/tests/server/test_http_output.py:27: note:      Subclass:
python/tests/server/test_http_output.py:27: note:          def predict(self) -> Path
python/tests/server/test_http_output.py:44: error: Function is missing a return type annotation
python/tests/server/test_http_output.py:46: error: Signature of "predict" incompatible with supertype "BasePredictor"
python/tests/server/test_http_output.py:46: note:      Superclass:
python/tests/server/test_http_output.py:46: note:          def predict(self, **kwargs: Any) -> Any
python/tests/server/test_http_output.py:46: note:      Subclass:
python/tests/server/test_http_output.py:46: note:          def predict(self) -> Path
python/tests/server/test_http_output.py:73: error: Function is missing a return type annotation
python/tests/server/test_http_output.py:75: error: Signature of "predict" incompatible with supertype "BasePredictor"
python/tests/server/test_http_output.py:75: note:      Superclass:
python/tests/server/test_http_output.py:75: note:          def predict(self, **kwargs: Any) -> Any
python/tests/server/test_http_output.py:75: note:      Subclass:
python/tests/server/test_http_output.py:75: note:          def predict(self) -> File
python/tests/server/test_http_output.py:76: error: Incompatible return value type (got "StringIO", expected "File")
python/tests/server/test_http_output.py:88: error: Function is missing a return type annotation
python/tests/server/test_http_output.py:90: error: Signature of "predict" incompatible with supertype "BasePredictor"
python/tests/server/test_http_output.py:90: note:      Superclass:
python/tests/server/test_http_output.py:90: note:          def predict(self, **kwargs: Any) -> Any
python/tests/server/test_http_output.py:90: note:      Subclass:
python/tests/server/test_http_output.py:90: note:          def predict(self) -> File
python/tests/server/test_http_output.py:93: error: Incompatible return value type (got "StringIO", expected "File")
python/tests/server/test_http_output.py:113: error: Function is missing a return type annotation
python/tests/server/test_http_output.py:115: error: Signature of "predict" incompatible with supertype "BasePredictor"
python/tests/server/test_http_output.py:115: note:      Superclass:
python/tests/server/test_http_output.py:115: note:          def predict(self, **kwargs: Any) -> Any
python/tests/server/test_http_output.py:115: note:      Subclass:
python/tests/server/test_http_output.py:115: note:          def predict(self) -> floating[_64Bit]
python/tests/server/test_http_output.py:124: error: Function is missing a return type annotation
python/tests/server/test_http_output.py:130: error: Signature of "predict" incompatible with supertype "BasePredictor"
python/tests/server/test_http_output.py:130: note:      Superclass:
python/tests/server/test_http_output.py:130: note:          def predict(self, **kwargs: Any) -> Any
python/tests/server/test_http_output.py:130: note:      Subclass:
python/tests/server/test_http_output.py:130: note:          def predict(self) -> Output
python/tests/server/test_http_input.py:5: error: Skipping analyzing "PIL": module is installed, but missing library stubs or py.typed marker
python/tests/server/test_http_input.py:13: error: Function is missing a return type annotation
python/tests/server/test_http_input.py:15: error: Signature of "predict" incompatible with supertype "BasePredictor"
python/tests/server/test_http_input.py:15: note:      Superclass:
python/tests/server/test_http_input.py:15: note:          def predict(self, **kwargs: Any) -> Any
python/tests/server/test_http_input.py:15: note:      Subclass:
python/tests/server/test_http_input.py:15: note:          def predict(self) -> str
python/tests/server/test_http_input.py:24: error: Function is missing a return type annotation
python/tests/server/test_http_input.py:26: error: Signature of "predict" incompatible with supertype "BasePredictor"
python/tests/server/test_http_input.py:26: note:      Superclass:
python/tests/server/test_http_input.py:26: note: (Skipping most remaining errors due to unresolved imports or missing stubs; fix these first)
python/tests/server/test_http_input.py:159: error: Name "test_file_input_with_http_url" already defined on line 103
Found 101 errors in 12 files (checked 21 source files)
make: *** [lint] Error 1
```

A particularly fun one is that the way we tell users to override `predict()` violates the Liskov substitution principle and I'm not sure how to do this in a way that makes those types that we tell everyone to define actually valid type definitions.

### CODE REVIEW

**Review by zeke — APPROVED** [review]
Looks good to me. 🧹 

I wonder if we could add a test that would prevent the number of infractions from going up in subsequent PRs. That might help us whittle it down over time.

**Comment by bfirsh** [comment]
It shouldn't take more than an hour or two of work to get this down to 0 (minus the tests, maybe) so maybe we can just jump straight to enforcing this properly.

File: requirements-dev.txt [hunk 1]
```diff
@@ -7,6 +7,8 @@ PyYAML==5.4.1
 redis==4.1.0
 requests==2.25.1
 responses==0.16.0
+types-requests==2.25.1
```
[the remark this answers is not in the document]
Comment (bfirsh) [reply requirements-dev.txt:10]: yes! thank you!

**Comment by bfirsh** [comment]
OK nevermind I fixed all those issues in the OP! Ready for another review.

I haven't added checking to the tests for the sake of sanity. It'll still show up as errors in VSCode for our information.

**Review by zeke — APPROVED** [review]
✨ 🧹

**Pull request merged**